cpp/src/arrow/python/datetime.cc

```diff
@@ -470,6 +474,28 @@ Result<std::string> TzinfoToString(PyObject* tzinfo) {
     return result;
   }
 
+  // try to look up key attribute
```

Comment:
Maybe add a comment here that this check is explicitly for zoneinfo objects, and below the same for dateutil